### PR TITLE
CORDA-676 Eager cordapp schemas

### PR DESCRIPTION
**Pull request by andr3ej**

### CODE REVIEW

**Comment by rick-r3** [comment]
`AbstractNode` looks correct to me.  Just going to look at the rest of it now...

File: client/rpc/src/integration-test/java/net/corda/client/rpc/CordaRPCJavaClientTest.java [hunk 1]
```diff
@@ -50,9 +50,8 @@ private void login(String username, String password) {
     @Before
     public void setUp() throws ExecutionException, InterruptedException {
         setCordappPackages("net.corda.finance.contracts");
-        CordaFuture<StartedNode<Node>> nodeFuture = startNotaryNode(getALICE().getName(), singletonList(rpcUser), true);
+        CordaFuture<StartedNode<Node>> nodeFuture = startNotaryNode(getALICE().getName(), singletonList(rpcUser), true, Collections.singleton(CashSchemaV1.INSTANCE));
```
Comment (rick-r3) [inline client/rpc/src/integration-test/java/net/corda/client/rpc/CordaRPCJavaClientTest.java]: Hmmm..... this is intriguing.  I don't think we should have to specify the custom schemas here.  If `setCordappPackages("net.corda.finance.contracts")` was `setCordappPackages("net.corda.finance")` then they should auto deploy.... I think?!  The schemas live in a package up from `contracts`.  Can you try it and see if it eliminates the need?

There's a PR to do something with `setCordappPackages` too though I think.  Maybe this flies in the face of that?

File: node/src/main/kotlin/net/corda/node/utilities/CordaPersistence.kt [hunk 2]
```diff
@@ -32,12 +30,7 @@ class CordaPersistence(var dataSource: HikariDataSource, private val schemaServi
             HibernateConfiguration(schemaService, databaseProperties, createIdentityService)
         }
     }
-
-    val entityManagerFactory: SessionFactory by lazy {
-        transaction {
```
Comment (Clintonio) [inline node/src/main/kotlin/net/corda/node/utilities/CordaPersistence.kt]: Is it okay to remove this transaction?

File: node/src/main/kotlin/net/corda/node/internal/cordapp/CordappProviderImpl.kt [hunk 3]
```diff
@@ -9,12 +9,15 @@ import net.corda.core.cordapp.CordappContext
 import net.corda.core.cordapp.CordappProvider
 import net.corda.core.node.services.AttachmentId
 import net.corda.core.serialization.SingletonSerializeAsToken
-import java.net.URLClassLoader
+
+interface CordappProviderInternal : CordappProvider {
```
Comment (Clintonio) [inline node/src/main/kotlin/net/corda/node/internal/cordapp/CordappProviderImpl.kt]: Please split to a separate file.

File: client/rpc/src/integration-test/java/net/corda/client/rpc/CordaRPCJavaClientTest.java (hunk 1 above)
Comment (Clintonio) [reply client/rpc/src/integration-test/java/net/corda/client/rpc/CordaRPCJavaClientTest.java]: I am not sure how the auto-JARing works with nested folders but yes, if it works as I'd like that should work. Though this does make everything a lot more explict instead of stringly typed which I prefer. Perhaps instead @andr3ej add a ``CordappBuilder`` list that allows you to specify any custom classes for cordapp loading instead of the current mess of setCordappPackages. We'll also need to figure out if anyone else is working in the area because I asked @andr3ej to look at it.

File: node/src/main/kotlin/net/corda/node/utilities/CordaPersistence.kt (hunk 2 above)
Comment (andr3ej) [reply node/src/main/kotlin/net/corda/node/utilities/CordaPersistence.kt]: Now that sessionFactoryForRegisteredSchemas is a val, it's created when HibernateConfiguration is constructed, which already has its own transaction a few lines up.

File: node/src/main/kotlin/net/corda/node/internal/cordapp/CordappProviderImpl.kt (hunk 3 above)
Comment (andr3ej) [reply node/src/main/kotlin/net/corda/node/internal/cordapp/CordappProviderImpl.kt]: Done

File: client/rpc/src/integration-test/java/net/corda/client/rpc/CordaRPCJavaClientTest.java (hunk 1 above)
Comment (andr3ej) [reply client/rpc/src/integration-test/java/net/corda/client/rpc/CordaRPCJavaClientTest.java]: e.g. CordaRPCClientTest.\`get cash balances` works without customSchemas if I add net.corda.finance.schemas as a cordapp package, also works if net.corda.finance is the only cordapp package.

Comment (rick-r3) [reply client/rpc/src/integration-test/java/net/corda/client/rpc/CordaRPCJavaClientTest.java]: Sounds promising.  In that case, can we eliminate all uses of `customSchemas` and not introduce the constructor param at all?  Or is there a counter example?

**Comment by andr3ej** [comment]
Working on it now.

**Comment by andr3ej** [comment]
@rick-r3 @Clintonio this is ready to be reviewed again, customSchemas is gone and a couple of changes were necessary in cordapploader/cordappprovider.

File: core/src/test/kotlin/net/corda/core/internal/InternalUtilsTest.kt [hunk 4]
```diff
@@ -61,6 +61,7 @@ class InternalUtilsTest {
         assertArrayEquals(intArrayOf(1, 2, 3, 4), (1 until 5).stream().toArray())
         assertArrayEquals(intArrayOf(1, 3), (1..4 step 2).stream().toArray())
         assertArrayEquals(intArrayOf(1, 3), (1..3 step 2).stream().toArray())
+        @Suppress("EmptyRange") // It's supposed to be empty.
```
Comment (andr3ej) [inline core/src/test/kotlin/net/corda/core/internal/InternalUtilsTest.kt]: fix unrelated warning

File: node/src/main/kotlin/net/corda/node/internal/Node.kt [hunk 5]
```diff
@@ -99,6 +99,7 @@ open class Node(override val configuration: FullNodeConfiguration,
     }
 
     override val log: Logger get() = logger
+    override val configuration get() = super.configuration as FullNodeConfiguration // Necessary to avoid init order NPE.
```
Comment (andr3ej) [inline node/src/main/kotlin/net/corda/node/internal/Node.kt]: override val creates a new field i think, which isn't ready yet when superclass wants to use it

File: node/src/main/kotlin/net/corda/node/internal/cordapp/CordappLoader.kt [hunk 6]
```diff
@@ -97,20 +97,22 @@ class CordappLoader private constructor(private val cordappJarPaths: List<URL>)
...
-        private fun createScanPackage(scanPackage: String): List<URL> {
+        private fun createScanPackage(scanPackage: String): List<RestrictedURL> {
             val resource = scanPackage.replace('.', '/')
             return this::class.java.classLoader.getResources(resource)
                     .asSequence()
                     .map { path ->
                         if (path.protocol == "jar") {
-                            (path.openConnection() as JarURLConnection).jarFileURL.toURI()
+                            // When running tests from gradle this may be a corda module jar, so restrict to scanPackage:
+                            RestrictedURL((path.openConnection() as JarURLConnection).jarFileURL, scanPackage)
```
Comment (andr3ej) [inline node/src/main/kotlin/net/corda/node/internal/cordapp/CordappLoader.kt]: this fixes one of the VaultQueryTests

File: node/src/main/kotlin/net/corda/node/internal/cordapp/CordappProviderImpl.kt [hunk 7]
```diff
@@ -34,28 +33,19 @@ open class CordappProviderImpl(private val cordappLoader: CordappLoader) : Singl
...
      */
-    fun getCordappAttachmentId(cordapp: Cordapp): SecureHash? = cordappAttachments.inverse().get(cordapp)
+    fun getCordappAttachmentId(cordapp: Cordapp): SecureHash? = cordappAttachments.inverse().get(cordapp.jarPath)
 
-    private fun loadContractsIntoAttachmentStore(attachmentStorage: AttachmentStorage): Map<SecureHash, Cordapp> {
-        val cordappsWithAttachments = cordapps.filter { !it.contractClassNames.isEmpty() }
-        val attachmentIds = cordappsWithAttachments.map { it.jarPath.openStream().use { attachmentStorage.importAttachment(it) } }
+    private fun loadContractsIntoAttachmentStore(attachmentStorage: AttachmentStorage): Map<SecureHash, URL> {
+        val cordappsWithAttachments = cordapps.filter { !it.contractClassNames.isEmpty() }.map { it.jarPath }
+        val attachmentIds = cordappsWithAttachments.map { it.openStream().use { attachmentStorage.importAttachment(it) } }
```
Comment (andr3ej) [inline node/src/main/kotlin/net/corda/node/internal/cordapp/CordappProviderImpl.kt]: cordapp equality is too strict for TraderDemoTest

File: node/src/main/kotlin/net/corda/node/utilities/CordaPersistence.kt [hunk 8]
```diff
@@ -103,7 +96,7 @@ class CordaPersistence(var dataSource: HikariDataSource, private val schemaServi
     }
 }
 
-fun configureDatabase(dataSourceProperties: Properties, databaseProperties: Properties?, schemaService: SchemaService = NodeSchemaService(), createIdentityService: () -> IdentityService): CordaPersistence {
+fun configureDatabase(dataSourceProperties: Properties, databaseProperties: Properties?, createIdentityService: () -> IdentityService, schemaService: SchemaService = NodeSchemaService(null)): CordaPersistence {
```
Comment (andr3ej) [inline node/src/main/kotlin/net/corda/node/utilities/CordaPersistence.kt]: swapped as createIdentityService is the more popular thing to pass in

File: node/src/test/kotlin/net/corda/node/internal/cordapp/CordappProviderImplTests.kt [hunk 9]
```diff
@@ -34,17 +33,14 @@ class CordappProviderImplTests {
...
-
+        val provider = CordappProviderImpl(loader, attachmentStore)
         Assert.assertNull(provider.getCordappAttachmentId(provider.cordapps.first()))
     }
 
     @Test
     fun `test that we find a cordapp class that is loaded into the store`() {
         val loader = CordappLoader.createDevMode(listOf(isolatedJAR))
-        val provider = CordappProviderImpl(loader)
+        val provider = CordappProviderImpl(loader, mock())
```
Comment (andr3ej) [inline node/src/test/kotlin/net/corda/node/internal/cordapp/CordappProviderImplTests.kt]: we should use mocks like this a lot more, not necessarily mockito though

File: node/src/test/kotlin/net/corda/node/services/persistence/HibernateConfigurationTest.kt [hunk 10]
```diff
@@ -95,13 +95,13 @@ class HibernateConfigurationTest : TestDependencyInjectionBase() {
...
         setUpDb()
-
-        val customSchemas = setOf(VaultSchemaV1, CashSchemaV1, SampleCashSchemaV2, SampleCashSchemaV3)
-        sessionFactory = hibernateConfig.sessionFactoryForSchemas(*customSchemas.toTypedArray())
+        sessionFactory = sessionFactoryForSchemas(VaultSchemaV1, CashSchemaV1, SampleCashSchemaV2, SampleCashSchemaV3)
         entityManager = sessionFactory.createEntityManager()
         criteriaBuilder = sessionFactory.criteriaBuilder
     }
 
+    private fun sessionFactoryForSchemas(vararg schemas: MappedSchema) = hibernateConfig.sessionFactoryForSchemas(schemas.toSet())
```
Comment (andr3ej) [inline node/src/test/kotlin/net/corda/node/services/persistence/HibernateConfigurationTest.kt]: a lot of api only existed to accommodate this test, so i've migrated the complexity here

File: node/src/test/kotlin/net/corda/node/services/vault/VaultQueryTests.kt [hunk 11]
```diff
@@ -45,10 +46,9 @@ import java.time.temporal.ChronoUnit
 import java.util.*
 
 class VaultQueryTests : TestDependencyInjectionBase() {
-    companion object {
-        private val cordappPackages = listOf("net.corda.testing.contracts", "net.corda.finance.contracts")
-    }
-
+    private val cordappPackages = setOf(
+            "net.corda.testing.contracts", "net.corda.finance.contracts",
+            CashSchemaV1::class.packageName, CommercialPaperSchemaV1::class.packageName, DummyLinearStateSchemaV1::class.packageName).toMutableList()
```
Comment (andr3ej) [inline node/src/test/kotlin/net/corda/node/services/vault/VaultQueryTests.kt]: 2 of these packages are the same, so i've used a temporary set to make the test below work

File: samples/network-visualiser/src/main/kotlin/net/corda/netmap/simulation/Simulation.kt [hunk 12]
```diff
@@ -270,9 +269,3 @@ abstract class Simulation(val networkSendManuallyPumped: Boolean,
         mockNet.stopNodes()
     }
 }
-
-/**
- * Helper function for verifying that a service info contains the given type of advertised service. For non-simulation cases
- * this is a configuration matter rather than implementation.
- */
-fun Iterable<ServiceInfo>.containsType(type: ServiceType) = any { it.type == type }
```
Comment (andr3ej) [inline samples/network-visualiser/src/main/kotlin/net/corda/netmap/simulation/Simulation.kt]: unrelated unused code

File: testing/node-driver/src/main/kotlin/net/corda/testing/node/MockServices.kt [hunk 13]
```diff
@@ -101,24 +100,22 @@ open class MockServices(
...
          *
          * @return a pair where the first element is the instance of [CordaPersistence] and the second is [MockServices].
          */
         @JvmStatic
-        fun makeTestDatabaseAndMockServices(customSchemas: Set<MappedSchema> = emptySet(),
-                                            keys: List<KeyPair> = listOf(MEGA_CORP_KEY),
+        fun makeTestDatabaseAndMockServices(keys: List<KeyPair> = listOf(MEGA_CORP_KEY),
                                             createIdentityService: () -> IdentityService = { makeTestIdentityService() },
                                             cordappPackages: List<String> = emptyList()): Pair<CordaPersistence, MockServices> {
+            val cordappLoader = CordappLoader.createWithTestPackages(cordappPackages)
```
Comment (andr3ej) [inline testing/node-driver/src/main/kotlin/net/corda/testing/node/MockServices.kt]: NodeSchemaService now takes a CordappLoader so i only have to create one of them in this use case
